src/uart/mod.rs

```diff
@@ -0,0 +1,81 @@
+//! Using the UART
+use crate::error::{NegativeErrorExt, NumericError};
+use embedded_io;
+use embedded_io::ErrorKind;
+use riot_sys::libc::c_void;
+use riot_sys::uart_init;
+use riot_sys::uart_t;
+use riot_sys::uart_write;
+
+#[derive(Debug)]
+pub struct UARTDevice {
+    dev: uart_t,
+}
+
+#[derive(Debug)]
+pub struct Error(NumericError);
+
+impl From<NumericError> for Error {
+    fn from(err: NumericError) -> Self {
+        Self(err)
+    }
+}
+
+impl embedded_io::Error for Error {
+    fn kind(&self) -> ErrorKind {
+        match -self.0.number() as _ {
+            riot_sys::ENODEV => ErrorKind::NotFound,
+            riot_sys::ENOTSUP => ErrorKind::Unsupported,
+            _ => ErrorKind::Other,
+        }
+    }
+}
+
+impl embedded_io::ErrorType for UARTDevice {
+    type Error = Error;
+}
+
+impl UARTDevice {
+    /// Create a new UARTDevice from a RIOT descriptor
+    pub fn new(dev: uart_t) -> Self {
+        UARTDevice { dev }
+    }
+
+    pub fn init(
+        &mut self,
+        baudrate: u32,
+        callback: Option<unsafe extern "C" fn(ctx: *mut c_void, byte: u8)>,
+        arg: Option<*mut c_void>,
```

Comment:
```suggestion
pub fn init<CTX_TYPE>(
        &mut self,
        baudrate: u32,
        callback: Option<fn(ctx: &mut CTX_TYPE, byte: u8)>,
        arg: Option<&mut CTX_TYPE>,
```

This would be much nicer...but I don't know how to do it.

Reply:
You're on the right track here.

What will help you in the right direction is if you try to make the function not an fn, but an `FnMut()` -- after all, we can be sure that interrupt calls don't preempt themselves, and this will allow the closure to own things.

What you'll then find is that not only the CTX_TYPE but the whole Callback becomes a type, and you'll have to take ownership of the function and store it somewhere.

The place to store it is in Self, which propagates the type into Self – looks weird, and `.new()` will be underspecified type-wise – but is that a problem, when type inference can kick in at init?

Reply:
After banging my head against the keyboard once more, I came to the conclusion that, after my educated attempts failed, I entered the area of try & error based programming. Something I would rather not do.

The compiler either stops me with lifetime struggles (the user provided `FnMut` not living long enough) or when I wrap it in an option or Box, that I can't cast a thin pointer to a thick pointer within the callback.

Reply:
With how the callback can live even after the concrete &mut self is dropped, the FnMut would likely also need to be 'static, and that's a legitimate restriction (still better than a `fn()` that is always static). Does that help?

Reply:
I am sorry chrysn, I think I am missing to many rust basics to figure this out :/